src/main/default/classes/FORM_PermissionValidator.cls

```diff
@@ -0,0 +1,146 @@
+    * @description Method for validating CRUD/FLS permissions for a template
+    * @param template - the template to use for CRUD/FLS validation
+    * @param accessLevels - a collection of CRUD/FLS validations to perform on the template (i.e. CREATE/UPDATE/READ)
+    */
+    public FORM_Template validateTemplate() {
+        Map<String, Set<String>> permissionErrorsBySObject = new Map<String, Set<String>>();
+        Map<String, Set<String>> fieldMappingsByTargetObject = new Map<String, Set<String>>();
+
+        BDI_FieldMappingSet fieldMappings =  BDI_MappingServiceAdvanced.getInstance(true).fieldMappingSet;
+        for (FORM_Section section : template.layout.sections) {
+            for (FORM_Element element : section.elements) {
+                for (String fieldDevName : element.dataImportFieldMappingDevNames) {
+                    if(fieldMappings.fieldMappingByDevName.containsKey(fieldDevName)) {
+                        BDI_FieldMapping fieldMapping = fieldMappings.fieldMappingByDevName.get(fieldDevName);
+
+                        String targetObjDevName = fieldMapping.Target_Object_API_Name;
+                        String targetFieldDevName = fieldMapping.Target_Field_API_Name;
+
+                        if(!isValid(targetObjDevName, accessLevels)) {
+                            permissionErrorsBySObject.put(targetObjDevName, new Set<String>{OBJECT_CRUD_ERROR});
+                        }
+                        
+                        if(!isValid(targetObjDevName, targetFieldDevName, accessLevels)) {
+                            if(permissionErrorsBySObject.containsKey(targetObjDevName)) {
+                                permissionErrorsBySObject.get(targetObjDevName).add(targetFieldDevName);
+                            } else {
+                                permissionErrorsBySObject.put(targetObjDevName, new Set<String>{targetFieldDevName});
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        handlePermissionsErrors(permissionErrorsBySObject);
+        return template;
+    }
+
+    /*******************************************************************************************************
+    * @description Perform CRUD validation on an SObject
+    * @param objDevName - the API name of the SObject to validate
+    * @param accessLevels - the levels of access to validate on the SObject
+    */
+    private Boolean isValid(String objDevName, Set<AccessLevel> accessLevels) {
+
+        SObjectType objType = ((SObject) Type.forName(objDevName).newInstance()).getSObjectType();
+
+        if((accessLevels.contains(AccessLevel.VALIDATE_CREATE)
+            && !permissionsService.canCreate(objType))
+            || (accessLevels.contains(AccessLevel.VALIDATE_READ)
+            && !permissionsService.canRead(objType))
+            || (accessLevels.contains(AccessLevel.VALIDATE_UPDATE)
```

Comment:
Be wary of the some of the gotchas with update - CampaignId on CampaignMember fails isUpdateable() even though it's a lookup, and many master-detail relations are un-reparentable and so fail this check as well; however these frequently might show up on a form to add campaign members, Account Soft credits, etc. FLS perms service has some special exceptions for these cases I believe.

Reply:
good call. fixed that.